### PR TITLE
BAVL-213 changes to support the handling of multiple appointment matchs on cancel/delete in external services.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/activitiesappointments/ActivitiesAppointmentsClient.kt [hunk 1]
```diff
@@ -73,7 +73,7 @@ class ActivitiesAppointmentsClient(private val activitiesAppointmentsApiWebClien
       log.info("A&A CLIENT: query params - prisonCode=$prisonCode, prisonerNumber=$prisonerNumber, onDate=$onDate, locationIds=${locationIds.toList()}")
       getPrisonersAppointments(prisonCode, prisonerNumber, onDate)
         .also { log.info("A&A CLIENT: matches pre-location filter: $it") }
-        .filter { locationIds.toList().contains(it.internalLocation?.id) }
+        .filter { locationIds.toList().contains(it.internalLocation?.id) && it.category.code == VIDEO_LINK_BOOKING }
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/activitiesappointments/ActivitiesAppointmentsClient.kt:76]: extra check to ensure it is a video link booking

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonapi/PrisonApiClient.kt [hunk 2]
```diff
@@ -64,7 +64,7 @@ class PrisonApiClient(private val prisonApiWebClient: WebClient) {
       log.info("PRISON-API CLIENT: query params - prisonCode=$prisonCode, prisonerNumber=$prisonerNumber, onDate=$onDate, locationIds=${locationIds.toList()}")
       getPrisonersAppointments(prisonCode, prisonerNumber, onDate)
         .also { log.info("PRISON-API CLIENT: matches pre-location filter: $it") }
-        .filter { locationIds.contains(it.locationId) }
+        .filter { locationIds.contains(it.locationId) && it.event == VIDEO_LINK_BOOKING }
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonapi/PrisonApiClient.kt:67]: extra check to ensure it is a video link booking

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/events/ManageExternalAppointmentsService.kt [hunk 3]
```diff
@@ -108,15 +110,21 @@ class ManageExternalAppointmentsService(
     )
   }
 
-  private fun Collection<AppointmentSearchResult>.findMatching(appointment: PrisonAppointment) =
-    singleOrNull {
+  private fun Collection<AppointmentSearchResult>.findMatchingActivitiesAndAppointments(appointment: PrisonAppointment): List<AppointmentSearchResult> =
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/events/ManageExternalAppointmentsService.kt:113]: There could potentially be multiple matches e.g. pre, main and post `singleOrNull` would reject this (at least that is what I think is wrong).

Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/events/ManageExternalAppointmentsService.kt:113]: Sounds a likely candidate.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonapi/PrisonApiClient.kt (hunk 2 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonapi/PrisonApiClient.kt:67]: Possibly yes.  Either way we have to assume there could be more than one.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/wiremock/ActivitiesAppointmentsApiMockServer.kt [hunk 4]
```diff
@@ -119,7 +120,7 @@ class ActivitiesAppointmentsApiMockServer : MockServer(8089) {
         appointmentSeriesId = 1,
         appointmentName = "appointment name",
         attendees = listOf(AppointmentAttendeeSearchResult(1, prisonerNumber, 1)),
-        category = AppointmentCategorySummary("VLB", "video link booking"),
+        category = AppointmentCategorySummary(locationType, "video link booking"),
         inCell = false,
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/wiremock/ActivitiesAppointmentsApiMockServer.kt:124]: Is this right? LocationType going into category?

**Review by timharrison-moj — APPROVED** [review]
LGTM.
Small question.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/wiremock/ActivitiesAppointmentsApiMockServer.kt (hunk 4 above)
Comment (matt-crowson-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/wiremock/ActivitiesAppointmentsApiMockServer.kt:124]: I believe so yes.  Bare in mind I have just called it this in the test so it reads as location type when calling.

**Pull request merged**